### PR TITLE
[BEAM-6207] Added option to publish synthetic data to Kafka topic.

### PR DESCRIPTION
**Pull request by mwalenia**
SyntheticDataPubSubPublisher now accepts --kafkaBootstrapServerAddress option
which switches it to Kafka publishing mode instead of Google PubSub mode.
Key-Value pairs are published as strings.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by mwalenia** [comment]
Hi @kkucharc , could you look at it? Thanks!

**Comment by mwalenia** [comment]
One test failure is in Task :beam-runners-google-cloud-dataflow-java-legacy-worker:test, unrelated to these changes.
Second one is checkstyle in /src/runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/job/ReferenceRunnerJobServer.java:147
, there's a Javadoc missing - also unrelated

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 1]
```diff
@@ -73,6 +80,11 @@
     String getInsertionPipelineTopic();
 
     void setInsertionPipelineTopic(String topic);
+
+    @Description("Kafka server address (optional)")
```
Comment (kkucharc) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: I would keep the info about this being optional only in documentation. Lack of `@Validation.Required` is enough. 

Usually when something is not required it has `@Default.` value. But as long as we treat this pipeline option as condition whether we use Kafka or PubSub, then it can leave it as it is now or change to default value `""` and then change the condition in the 98 line. WDYT?

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 2]
```diff
@@ -83,14 +95,41 @@ public static void main(String[] args) throws IOException {
 
     Pipeline pipeline = Pipeline.create(options);
 
-    pipeline
-        .apply("Read synthetic data", Read.from(new SyntheticBoundedSource(sourceOptions)))
-        .apply("Map to PubSub messages", MapElements.via(new MapBytesToPubSubMessage()))
-        .apply("Write to PubSub", PubsubIO.writeMessages().to(options.getInsertionPipelineTopic()));
-
+    if (options.getKafkaBootstrapServerAddress() != null) {
```
Comment (kkucharc) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: Personally, I would recommend to split those two pipelines into separate files. 

@lgajowy WDYT?

**Comment by kkucharc** [comment]
Run Java PreCommit

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java (hunk 1 above)
Comment (mwalenia) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: That's a good idea, done.

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 3]
```diff
@@ -73,6 +81,12 @@
     String getInsertionPipelineTopic();
 
     void setInsertionPipelineTopic(String topic);
+
+    @Description("Kafka server address")
+    @Default.String("")
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: Could you remove this default and do a null-check instead?

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 4]
```diff
@@ -50,9 +54,13 @@
  * <pre>
  *  ./gradlew :beam-sdks-java-load-tests:run -PloadTest.args='
  *    --insertionPipelineTopic=TOPIC_NAME
+ *    --kafkaBootstrapServerAddress=SERVER_ADDRESS
  *    --sourceOptions={"numRecords":1000,...}'
  *    -PloadTest.mainClass="org.apache.beam.sdk.loadtests.SyntheticDataPubSubPublisher"
  *  </pre>
+ *
+ * Parameter kafkaBootstrapServerAddress is optional. If provided, pipeline topic will be treated as
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: I think we can do it even better. This pipeline could be designed in such a way that allows publishing to every sink type at the same time. So having: 

 - kafka related options & pubsub related options we can publish to both Kafka and Pubsub 
 - kafka related options: publish only to Kafka
 - pubsub related options: publish only to PubSub

WDYT?

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 5]
```diff
@@ -83,14 +97,41 @@ public static void main(String[] args) throws IOException {
 
     Pipeline pipeline = Pipeline.create(options);
 
-    pipeline
-        .apply("Read synthetic data", Read.from(new SyntheticBoundedSource(sourceOptions)))
-        .apply("Map to PubSub messages", MapElements.via(new MapBytesToPubSubMessage()))
-        .apply("Write to PubSub", PubsubIO.writeMessages().to(options.getInsertionPipelineTopic()));
-
+    if (!options.getKafkaBootstrapServerAddress().isEmpty()) {
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: Those sinks are somewhat independent and we could let them stay this way. We can code it like this (including the division of the methods):

```
if (<options for kafka are present) {
      writeToKafka(collection);

if (<options for pubsub are present>)
      writeToPubsub(collection);
}
```

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 6]
```diff
@@ -83,14 +97,41 @@ public static void main(String[] args) throws IOException {
...
+              "Write to PubSub", PubsubIO.writeMessages().to(options.getInsertionPipelineTopic()));
+    }
     pipeline.run().waitUntilFinish();
   }
 
+  private static class MapKVToString extends SimpleFunction<KV<byte[], byte[]>, String> {
+    @Override
+    public String apply(KV<byte[], byte[]> input) {
+      StringBuilder stringBuilder =
+          new StringBuilder()
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: Hint: an "oneliner" for this can be done using `String.format("{%s,%s}", Arrays.toString(input.getKey()), Arrays.toString(input.getValue()))`

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java (hunk 2 above)
Comment (lgajowy) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: IMO it can stay as one file. After all, we can have multiple sinks in one pipeline and even publish to both at the same time if we want to (no idea that this is ever going to be a case, though...  :) ) . I'd split it like I suggested in the review comments below.

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java [hunk 7]
```diff
@@ -50,9 +54,13 @@
...
  *    --insertionPipelineTopic=TOPIC_NAME
+ *    --kafkaBootstrapServerAddress=SERVER_ADDRESS
  *    --sourceOptions={"numRecords":1000,...}'
  *    -PloadTest.mainClass="org.apache.beam.sdk.loadtests.SyntheticDataPubSubPublisher"
  *  </pre>
+ *
+ * Parameter kafkaBootstrapServerAddress is optional. If provided, pipeline topic will be treated as
+ * Kafka topic name and records will be published to Kafka instead of PubSub.
  */
 public class SyntheticDataPubSubPublisher {
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: Let's change this name to abstract from `PubSub`.

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java (hunk 3 above)
Comment (lgajowy) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: one last thing: we need to change the class name as it's not only about `PubSub` now :)

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java (hunk 4 above)
Comment (mwalenia) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: That's a good idea, implemented in 2nd commit

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java (hunk 5 above)
Comment (mwalenia) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPubSubPublisher.java]: implemented, thanks for the suggestion!

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java [hunk 8]
```diff
@@ -69,26 +80,64 @@
     void setSourceOptions(String sourceOptions);
 
     @Description("PubSub topic to publish to")
-    @Validation.Required
     String getInsertionPipelineTopic();
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java]: Can you change this to `getPubSubTopic` too? `InsertionPipelineTopic` is too generic now since we have the Kafka option too.

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java [hunk 9]
```diff
@@ -69,26 +80,64 @@
...
 
   public static void main(String[] args) throws IOException {
-    Options options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
+    options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
 
     SyntheticSourceOptions sourceOptions =
         SyntheticOptions.fromJsonString(options.getSourceOptions(), SyntheticSourceOptions.class);
 
     Pipeline pipeline = Pipeline.create(options);
+    PCollection<KV<byte[], byte[]>> syntheticData =
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java]: looks good!

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java [hunk 10]
```diff
@@ -50,15 +54,22 @@
  * <pre>
  *  ./gradlew :beam-sdks-java-load-tests:run -PloadTest.args='
  *    --insertionPipelineTopic=TOPIC_NAME
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java]: Please remember to change here too

File: sdks/java/io/synthetic/build.gradle [hunk 11]
```diff
@@ -29,11 +29,13 @@ dependencies {
   shadow library.java.jackson_annotations
   shadow library.java.jackson_databind
   shadow library.java.guava
+  shadow library.java.kafka_clients
```
Comment (lgajowy) [inline sdks/java/io/synthetic/build.gradle]: Why is this needed in `sdks/java/io/synthetic/build.gradle`? Shouldn't we add this in `sdks/java/testing/load-tests/build.gradle`?

File: sdks/java/io/synthetic/build.gradle [hunk 12]
```diff
@@ -29,11 +29,13 @@ dependencies {
...
 
   shadowTest library.java.vendored_guava_20_0
   testCompile library.java.junit
   testCompile library.java.hamcrest_core
   testCompile library.java.hamcrest_library
   shadow project(path: ":beam-sdks-java-core", configuration: "shadow")
   shadow project(path: ":beam-runners-direct-java", configuration: "shadow")
+  shadow project(path: ":beam-sdks-java-io-kafka", configuration: "shadow")
```
Comment (lgajowy) [inline sdks/java/io/synthetic/build.gradle]: (as the comment above)

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java [hunk 13]
```diff
@@ -50,15 +54,22 @@
...
+ *    --kafkaBootstrapServerAddress=SERVER_ADDRESS
+ *    --kafkaTopic=KAFKA_TOPIC_NAME
  *    --sourceOptions={"numRecords":1000,...}'
- *    -PloadTest.mainClass="org.apache.beam.sdk.loadtests.SyntheticDataPubSubPublisher"
+ *    -PloadTest.mainClass="org.apache.beam.sdk.loadtests.SyntheticDataPublisher"
  *  </pre>
+ *
+ * If parameters related to Kafka are provided, the publisher writes to Kafka. If both pubsub topic
+ * and Kafka params are present, records will be written to both sinks.
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java]: It would be even more clear if the sentence gives instructions for both Kafka and Pubsub, so I suggest: 

```
If parameters related to a specific sink are provided (Kafka or PubSub), the pipeline writes to the sink. Writing to both sinks is also acceptable.
``` 
(Or similar but including both runners)

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java [hunk 14]
```diff
@@ -50,15 +54,22 @@
  * <pre>
  *  ./gradlew :beam-sdks-java-load-tests:run -PloadTest.args='
  *    --insertionPipelineTopic=TOPIC_NAME
+ *    --kafkaBootstrapServerAddress=SERVER_ADDRESS
+ *    --kafkaTopic=KAFKA_TOPIC_NAME
  *    --sourceOptions={"numRecords":1000,...}'
- *    -PloadTest.mainClass="org.apache.beam.sdk.loadtests.SyntheticDataPubSubPublisher"
+ *    -PloadTest.mainClass="org.apache.beam.sdk.loadtests.SyntheticDataPublisher"
  *  </pre>
+ *
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java]: missing javadoc `<p>`

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java (hunk 8 above)
Comment (mwalenia) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/SyntheticDataPublisher.java]: Ok, done

**Comment by lgajowy** [comment]
LGTM, thanks! Merging.
